File: hard/2435. Paths in Matrix Whose Sum Is Divisible by K/solution.java
class Solution {
    public int numberOfPaths(int[][] grid, int k) {
        int m = grid.length, n = grid[0].length;
        int[][][] dp = new int[m][n][k];
        int mod = 1000000007;
        
        dp[0][0][grid[0][0] % k] = 1;
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                for (int r = 0; r < k; r++) {
                    if (i > 0) {
                        int newR = (r + grid[i][j]) % k;
                        dp[i][j][newR] = (dp[i][j][newR] + dp[i-1][j][r]) % mod;
                    }
                    if (j > 0) {
                        int newR = (r + grid[i][j]) % k;
                        dp[i][j][newR] = (dp[i][j][newR] + dp[i][j-1][r]) % mod;
                    }
                }
            }
        }
        
        return dp[m-1][n-1][0];
    }
}

File: hard/2435. Paths in Matrix Whose Sum Is Divisible by K/solution.rs
impl Solution {
    pub fn number_of_paths(grid: Vec<Vec<i32>>, k: i32) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        let k = k as usize;
        let mod_val = 1_000_000_007;
        
        let mut dp = vec![vec![vec![0; k]; n]; m];
        dp[0][0][(grid[0][0] as usize) % k] = 1;
        
        for i in 0..m {
            for j in 0..n {
                for r in 0..k {
                    if i > 0 {
                        let new_r = (r + grid[i][j] as usize) % k;
                        dp[i][j][new_r] = (dp[i][j][new_r] + dp[i-1][j][r]) % mod_val;
                    }
                    if j > 0 {
                        let new_r = (r + grid[i][j] as usize) % k;
                        dp[i][j][new_r] = (dp[i][j][new_r] + dp[i][j-1][r]) % mod_val;
                    }
                }
            }
        }
        
        dp[m-1][n-1][0]
    }
}